pkg/util/parameters.go

```diff
@@ -0,0 +1,180 @@
+package util
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	ackerr "github.com/aws-controllers-k8s/runtime/pkg/errors"
+)
+
+var (
+	errUnknownParameter      = fmt.Errorf("unknown parameter")
+	errUnmodifiableParameter = fmt.Errorf("parameter is not modifiable")
+)
+
+func NewErrUnknownParameter(name string) error {
+	// This is a terminal error because unless the user removes this parameter
+	// from their list of parameter overrides, we will not be able to get the
+	// resource into a synced state.
+	return ackerr.NewTerminalError(
+		fmt.Errorf("%w: %s", errUnknownParameter, name),
+	)
+}
+
+func NewErrUnmodifiableParameter(name string) error {
+	// This is a terminal error because unless the user removes this parameter
+	// from their list of parameter overrides, we will not be able to get the
+	// resource into a synced state.
+	return ackerr.NewTerminalError(
+		fmt.Errorf("%w: %s", errUnmodifiableParameter, name),
+	)
+}
+
+// ComputeParametersDelta compares two Parameter arrays and returns the new
+// parameters to add, to update and the parameter identifiers to delete
+func ComputeParametersDelta(
+	desired map[string]*string,
+	latest map[string]*string,
+) (map[string]*string, []string) {
+	toReset := []string{}
+	toModify := map[string]*string{}
+
+	for k, v := range desired {
+		if lv, found := latest[k]; !found {
+			toModify[k] = v
+		} else if !equalStrings(v, lv) {
+			toModify[k] = v
+		}
+	}
+	for k := range latest {
+		if _, found := desired[k]; !found {
+			toReset = append(toReset, k)
+		}
+	}
+	return toModify, toReset
+}
+
+// sliceStringChunks splits a supplied slice of string pointers into multiple
+// slices of string pointers of a given size.
+func SliceStringChunks(
+	input []string,
+	chunkSize int,
+) [][]string {
+	var chunks [][]string
+	for {
+		if len(input) == 0 {
+			break
+		}
+
+		if len(input) < chunkSize {
+			chunkSize = len(input)
+		}
+
+		chunks = append(chunks, input[0:chunkSize])
+		input = input[chunkSize:]
+	}
+
+	return chunks
+}
+
+// mapStringChunks splits a supplied map of string pointers into multiple
+// slices of maps of string pointers of a given size.
+func MapStringChunks(
+	input map[string]*string,
+	chunkSize int,
+) []map[string]*string {
+	var chunks []map[string]*string
+	chunk := map[string]*string{}
+	idx := 0
+	for k, v := range input {
+		if idx < chunkSize {
+			chunk[k] = v
+			idx++
+		} else {
+			// reset the chunker
+			chunks = append(chunks, chunk)
+			chunk = map[string]*string{}
+			idx = 0
+		}
+	}
+	chunks = append(chunks, chunk)
+
+	return chunks
+}
+
+// *** Caching
```

Comment:
nit: move all of this under a new file `caching.go`

Reply:
Or `param_cache.go`

Reply:
Done

(`parameter_cache.go`)